### ISSUE TITLE
Fixed: JUnit38ClassRunner does not recognise AssumptionViolationException (Assume)

### ISSUE THREAD

**Issue by nickl-** [junit-team/junit4#1490]
If you also ran into the problem where using JUnit3 style `TestSuite` running with `AllTests.class` or i.o.w. using compatibility mode JUnit38ClassRunner it is impossible to skip tests. As #1100 points out `Assume` is JUnit4 style and we should rewrite our tests if we don't want `AssumptionViolationException` to register as an error.  `@Ignore` doesn't seem to work either, didn't even attempt to apply it at runtime. 

The real issue though, is the fact that it is much easier to dynamically create JUnit3 `TestCase` classes then attempting to dynamically create JUnit4 `@Test` annotated class methods to accomplish the same. Since they are by nature dynamically created not having a means to dynamically skip these tests is a real catch 22.

If like me you find this unacceptable, perhaps this will do the trick:

```java

import java.lang.reflect.Method;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.junit.internal.runners.SuiteMethod;
import org.junit.runners.BlockJUnit4ClassRunner;
import org.junit.runners.model.FrameworkMethod;
import org.junit.runners.model.InitializationError;

import junit.framework.Test;
import junit.framework.TestCase;
import junit.framework.TestSuite;

public class AllTestsJUnit4Runner extends BlockJUnit4ClassRunner {
    private final Method method;
    private final TestSuite suite;

    public AllTestsJUnit4Runner(Class<?> klass) throws InitializationError {
        super(klass);
        try {
            suite = (TestSuite) SuiteMethod.testFromSuiteMethod(klass);
            method = TestCase.class.getDeclaredMethod("runBare");
        } catch (Throwable t) { throw new InitializationError(t); }
    }

    @Override
    protected List<FrameworkMethod> getChildren() {
        return Stream.concat(
                   super.getChildren().stream(), 
                   Collections.list(suite.tests()).stream()
                       .map(TestSuiteMethod::new))
               .collect(Collectors.toList());
    }

    class TestSuiteMethod extends FrameworkMethod {
        TestCase test;

        public TestSuiteMethod(Test test) {
            super(method);
            this.test = (TestCase) test;
        }

        @Override
        public Object invokeExplosively(final Object target, final Object... params)
                throws Throwable {
            test.runBare();
            return void.class;
        }

        @Override
        public String getName() {
            return test.getName();
        }
        
        @Override
        public boolean equals(Object obj) {
            if (!(obj instanceof TestSuiteMethod))
                return false;
            return test.equals(((TestSuiteMethod) obj).test); 
        }

        @Override
        public int hashCode() {
            return test.hashCode();
        }
    };   
}
```

Just replace the `@RunWith(AllTests.class)` annotation with `@RunWith(AllTestsJUnit4Runner.class)` and the `TestSuite` will be run complete with `BlockJUnit4ClassRunner` by processing each `TestCase` added to a `TestSuite` retrievable from the static `suite` method and running them as if they were JUnit4 style methods.

In addition the runner supports the following:
 * Will also run any JUnit4 style `@Test` annotated methods.
 * Respects JUnit3 style `setUp` and `tearDown` as expected.
 * Assigns the appropriate test name from each `TestCase`

@junit-team if you would like me to submit this as a PR I would gladly oblige...

**Comment by kcooney** [comment]
Thanks for raising this issue.

I don't know how the other maintainers feel but I am included not to accept a pull like this. We released JUnit 4 over ten years ago and haven't to my knowledge made any improvements to the JUnit 3 code in that time. We also just released JUnit 5. So TestCase is two generations old. Developers should not be using it for new tests.

If you want to dynamically create tests JUnit 5 has good support for that. If you cannot use JUnit 5 and want to dynamically create tests and add them to a suite in JUnit 4, use a Runner to dynamically create your test.
